dvc/checkout.py

```diff
@@ -315,7 +316,7 @@ def checkout(
         skip = True
     else:
         try:
-            check(cache, obj)
+            check(cache, obj, skip_tree=skip_tree_check)
```

Comment:
Depending on how we are doing checkout, `obj` may be a filtered memfs tree where we only need to checkout (link) the file entries, and don't care about the `.dir` cache being present

Reply:
This does look a bit odd though, but it is clear that the reason is that our `.dir`'s are just weird like that when it comes to filtered things. If we were to switch to git-like Tree objects, this flag would not be needed, right? Just wondering if there is a more elegant way to go about this right now.

Another interesting question is if we even need this `check()` call here. Like, previously we've done it because we used `hash_info`s here and didn't know if the objects actually exist or not. And these days we handle real objects here at this point, so maybe we can just skip this unnecessary check. Maybe I'm missing something...

Reply:
As long as we are always doing the correct `odb.save()` before `checkout` this check shouldn't be needed any more.

We could move it into `_checkout_file` and just check the individual file objects before linking if we want to be safe, but the dir-cache/tree obj check isn't needed at all.

Reply:
> We could move it into _checkout_file and just check the individual file objects before linking if we want to be safe, but the dir-cache/tree obj check isn't needed at all.

Are per-file checks needed though? Seems like that is also pretty safe. Though our linking logic might be not raising proper exceptions when a src is missing.

Reply:
I think we can get rid of the per-file checks then.

I'll double check what exceptions get raised when a src file is missing on link, but just to be clear, when a specific file is missing we should still checkout the rest of the file objects that do exist (and warn about the missing files), correct?

@efiop

Reply:
@pmrowla Didn't mean to expand the scope here too much, we could get back to this in a followup and live with the flag  in the mean time.

> I'll double check what exceptions get raised when a src file is missing on link, but just to be clear, when a specific file is missing we should still checkout the rest of the file objects that do exist (and warn about the missing files), correct?

Even though it doesn't match current behaviour (we skip the whole dir if it is incomplete, which is very questionable), it does make sense to checkout what we can, yes.

Reply:
Handling it in a follow-up is fine with me, and leaving this as-is with the extra flag for now will keep the current behavior since it still does the file checks before linking